test/hotspot/jtreg/compiler/c2/irTests/LShiftLNodeIdealizationTests.java

```diff
@@ -113,4 +125,99 @@ public long test7(long x) {
     public long test8(long x) {
         return ((x >>> 4L) & 0xFFL) << 8L;
     }
+
+    @DontCompile
+    public void assertDoubleShiftResult(long a) {
+        Asserts.assertEQ((a << 2L) << 3L, testDoubleShift1(a));
+        Asserts.assertEQ(a << 5L, testDoubleShift1(a));
+
+        Asserts.assertEQ(((a << 2L) << 3L) << 1L, testDoubleShift2(a));
+        Asserts.assertEQ(a << 6L, testDoubleShift2(a));
+
+        Asserts.assertEQ((a << 63L) << 1L, testDoubleShift3(a));
+        Asserts.assertEQ(0L, testDoubleShift3(a));
+
+        Asserts.assertEQ((a << 1L) << 63L, testDoubleShift4(a));
+        Asserts.assertEQ(0L, testDoubleShift4(a));
+
+        Asserts.assertEQ(((a << 62L) << 1L) << 1L, testDoubleShift5(a));
+        Asserts.assertEQ(0L, testDoubleShift5(a));
+
+        Asserts.assertEQ((a * 4L) << 3L, testDoubleShift6(a));
+        Asserts.assertEQ(a << 5L, testDoubleShift6(a));
+
+        Asserts.assertEQ((a << 3L) * 4L, testDoubleShift7(a));
+        Asserts.assertEQ(a << 5L, testDoubleShift7(a));
+
+        Asserts.assertEQ(a << 65L, testDoubleShift8(a));
+        Asserts.assertEQ(a << 1L, testDoubleShift8(a));
+
+        Asserts.assertEQ((a << 62L) << 3L, testDoubleShift9(a));
+        Asserts.assertEQ(0L, testDoubleShift9(a));
+    }
+
+    @Test
+    @IR(counts = {IRNode.LSHIFT, "1"})
+    // Checks (x << 2) << 3 => x << 5
+    public long testDoubleShift1(long x) {
+        return (x << 2L) << 3L;
+    }
+
+    @Test
+    @IR(counts = {IRNode.LSHIFT, "1"})
+    // Checks ((x << 2) << 3) << 1 => x << 6
+    public long testDoubleShift2(long x) {
+        return ((x << 2L) << 3L) << 1L;
+    }
+
+    @Test
+    @IR(failOn = {IRNode.LSHIFT})
+    // Checks (x << 63) << 1 => 0
+    public long testDoubleShift3(long x) {
+        return (x << 63L) << 1L;
+    }
+
+    @Test
+    @IR(failOn = {IRNode.LSHIFT})
+    // Checks (x << 1) << 63 => 0
+    public long testDoubleShift4(long x) {
+        return (x << 1L) << 63L;
+    }
+
+    @Test
+    @IR(failOn = {IRNode.LSHIFT})
+    // Checks ((x << 62) << 1) << 1 => 0
+    public long testDoubleShift5(long x) {
+        return ((x << 62L) << 1L) << 1L;
+    }
+
+    @Test
+    @IR(failOn = {IRNode.MUL})
+    @IR(counts = {IRNode.LSHIFT, "1"})
+    // Checks (x * 4) << 3 => x << 5
+    public long testDoubleShift6(long x) {
+        return (x * 4L) << 3L;
+    }
+
+    @Test
+    @IR(failOn = {IRNode.MUL})
+    @IR(counts = {IRNode.LSHIFT, "1"})
+    // Checks (x << 3) * 4 => x << 5
+    public long testDoubleShift7(long x) {
+        return (x << 3L) * 4L;
+    }
+
+    @Test
+    @IR(counts = {IRNode.LSHIFT, "1"})
+    // Checks x << 65 => x << 1
+    public long testDoubleShift8(long x) {
+        return x << 65L;
+    }
+
+    @Test
+    @IR(failOn = {IRNode.LSHIFT})
+    // Checks (x << 62) << 3 => 0
+    public long testDoubleShift9(long x) {
+        return (x << 62L) << 3L;
+    }
```

Comment:
I see that you have quite a few examples here with fixed constants. It would be good to extend this with random constants. You can do that with `static final` fields, as they are constant by the time we JIT compile.

```
private static final int CON0 = RANDOM.nextInt();
private static final int CON1 = RANDOM.nextInt();

    @Test
    public long test(int x) {
        return (x << CON0) << CON1;
    }
```
I would give you bonus points if you use `Generators.java`, because that produces more interesting constants.

Reply:
I can also do that, but I think using the constants is more valuable than only random numbers: most of random int or long won't have some corner properties I try to cover here. On top of that, with the comments, they help enumerating clearly different cases to look at. So, I'd be happy to add some randomization, but I strongly feel I should keep the constant ones as well.

Reply:
Absolutely, please keep the cases you already have, and add some randomized cases on top of it.

> most of random int or long won't have some corner properties I try to cover here

Right, that's why it is better to use `Generators.java` rather than just `Random`. With `Generators`, we make sure to generate more special values, such as powers of 2. And powers of 2 multiplication can for example be converted to shift, and that seems to be quite relevant here ;)

Reply:
I've tried something but... It seems too easy. With this way of doing CON0 and CON1, we only test 1 pair of constants at each run, no? I thought about a loop to cover ranges, but then, it's not constants for the compiler.

So... is that how am I supposed to do? Just one pair, and the number of run of the same test will provide with time some coverage of the parameter space?

Reply:
I think just one pair is good enough. This is more of a long-term test, that just fuzzes away every time its run. Such tests have a surprising coverage eventually.